pkg/preemptiontoleration/preemption_toleration.go

```diff
@@ -0,0 +1,416 @@
+		klog.V(5).Infof("Pod %v/%v is not eligible for more preemption.", preemptor.Namespace, preemptor.Name)
+		return "", nil
+	}
+
+	// 2) Find all preemption candidates.
+	candidates, err := pl.FindCandidates(ctx, state, preemptor, m)
+	if err != nil || len(candidates) == 0 {
+		return "", err
+	}
+
+	// 3) Interact with registered Extenders to filter out some candidates if needed.
+	candidates, err = dp.CallExtenders(ph.Extenders(), preemptor, nodeLister, candidates)
+	if err != nil {
+		return "", err
+	}
+
+	// 4) Find the best candidate.
+	bestCandidate := dp.SelectCandidate(candidates)
+	if bestCandidate == nil || len(bestCandidate.Name()) == 0 {
+		return "", nil
+	}
+
+	// 5) Perform preparation work before nominating the selected candidate.
+	if err := dp.PrepareCandidate(bestCandidate, pl.fh, cs, preemptor); err != nil {
+		return "", err
+	}
+
+	return bestCandidate.Name(), nil
+}
+
+// CanToleratePreemption evaluates whether the victimCandidate
+// pod can tolerate from preemption by the preemptor pod or not
+// by inspecting PriorityClass of victimCandidate pod.
+func CanToleratePreemption(
+	victimCandidate, preemptor *v1.Pod,
+	pcLister schedulinglisters.PriorityClassLister,
+	now time.Time,
+) (bool, error) {
+	preemptorPriority := corev1helpers.PodPriority(preemptor)
+
+	victimPriorityClass, err := pcLister.Get(victimCandidate.Spec.PriorityClassName)
+	if err != nil {
+		return false, err
+	}
+
+	pt, err := getCompletedPreemptionToleration(*victimPriorityClass)
+	if err != nil {
+		return false, nil
+	}
+
+	// check it can tolerate the preemption in terms of priority value
+	canTolerateOnPriorityValue := preemptorPriority < *pt.MinimumPreemptablePriority
+	if !canTolerateOnPriorityValue {
+		return canTolerateOnPriorityValue, nil
+	}
+
+	if pt.TolerationSeconds == nil {
+		return canTolerateOnPriorityValue, nil
+	}
+
+	// check it can tolerate the preemption in terms of toleration seconds
+	_, scheduledCondition := podutil.GetPodCondition(&victimCandidate.Status, v1.PodScheduled)
+	if scheduledCondition == nil {
+		// if victim is not scheduled, skip evaluating tolerationSeconds
+		return canTolerateOnPriorityValue, nil
+	}
+	canTolerateOnTolerationSeconds := (pt.TolerationSeconds != nil) &&
+		!now.After(scheduledCondition.LastTransitionTime.Time.Add(time.Duration(*pt.TolerationSeconds)*time.Second))
+
+	return canTolerateOnPriorityValue && canTolerateOnTolerationSeconds, nil
+}
+
+func getCompletedPreemptionToleration(
+	pc schedulingv1.PriorityClass,
+) (*config.PreemptionToleration, error) {
+	tolerationConfigRaw, ok := pc.Annotations[AnnotationKey]
+	if !ok {
+		return nil, nil
+	}
+
+	pt := &config.PreemptionToleration{}
+	_, _, err := configDecoder.Decode(([]byte)(tolerationConfigRaw), nil, pt)
+	if err != nil {
+		return nil, err
+	}
+
+	if pt.MinimumPreemptablePriority == nil {
+		pt.MinimumPreemptablePriority = pointer.Int32Ptr(pc.Value + 1)
+	}
+
```

Comment:
`TolerationSeconds: nil` means it can tolerate forever, and `TolerationSeconds: 0` means it can not tolerate forever.  So, completing `0` does not help here.